src/theory/arith/theory_arith.cpp

```diff
@@ -253,59 +302,29 @@ bool TheoryArith::collectModelInfo(TheoryModel* m,
 bool TheoryArith::collectModelValues(TheoryModel* m,
                                      const std::set<Node>& termSet)
 {
-  // get the model from the linear solver
-  std::map<Node, Node> arithModel;
-  d_internal->collectModelValues(termSet, arithModel);
-  // Double check that the model from the linear solver respects integer types,
-  // if it does not, add a branch and bound lemma. This typically should never
-  // be necessary, but is needed in rare cases.
-  bool addedLemma = false;
-  bool badAssignment = false;
-  for (const std::pair<const Node, Node>& p : arithModel)
+  if (Trace.isOn("arith::model"))
   {
-    if (p.first.getType().isInteger() && !p.second.getType().isInteger())
+    Trace("arith::model") << "arithmetic model after pruning" << std::endl;
+    for (const auto& p : d_arithModelCache)
     {
-      Assert(false) << "TheoryArithPrivate generated a bad model value for "
-                       "integer variable "
-                    << p.first << " : " << p.second;
-      // must branch and bound
-      TrustNode lem =
-          d_bab.branchIntegerVariable(p.first, p.second.getConst<Rational>());
-      if (d_im.trustedLemma(lem, InferenceId::ARITH_BB_LEMMA))
-      {
-        addedLemma = true;
-      }
-      badAssignment = true;
+      Trace("arith::model") << "\t" << p.first << " -> " << p.second << std::endl;
     }
   }
-  if (addedLemma)
-  {
-    // we had to add a branch and bound lemma since the linear solver assigned
-    // a non-integer value to an integer variable.
-    return false;
-  }
-  // this would imply that linear arithmetic's model failed to satisfy a branch
-  // and bound lemma
-  AlwaysAssert(!badAssignment)
-      << "Bad assignment from TheoryArithPrivate::collectModelValues, and no "
-         "branching lemma was sent";
 
-  // if non-linear is enabled, intercept the model, which may repair its values
-  if (d_nonlinearExtension != nullptr)
-  {
-    // Non-linear may repair values to satisfy non-linear constraints (see
-    // documentation for NonlinearExtension::interceptModel).
-    d_nonlinearExtension->interceptModel(arithModel, termSet);
-  }
   // We are now ready to assert the model.
-  for (const std::pair<const Node, Node>& p : arithModel)
+  for (const std::pair<const Node, Node>& p : d_arithModelCache)
   {
+    if (termSet.find(p.first) == termSet.end())
+    {
+      continue;
+    }
     // maps to constant of comparable type
     Assert(p.first.getType().isComparableTo(p.second.getType()));
     if (m->assertEquality(p.first, p.second, true))
     {
       continue;
     }
+    AlwaysAssert(false) << "A model equality could not be asserted: " << p.first << " == " << p.second << std::endl;
```

Comment:
This is making things harsher than previously.  Why can't we continue to fall back to the previous policy (in the block below this?)

Otherwise, I'd make this an `Assert` not `AlwaysAssert`.